# 6.1: Sourcing WildFires Dataset and Understanding Data
** **
## Contents:

1. Importing libraries and files


2. Previewing data, checking columns, and data types 


3. Data quality checks
      - 3-1 Duplicates     
      - 3-2 Missing values      
      - 3-3 Mixed data types
      
  
4. Data cleaning

    - 4-1 Renaming columns' names to keep names clear and consistent
    - 4-2 Missing values treatment
    - 4-3 Removing duplicates
    - 4-4 Dropping unnecessary columns
    - 4-5 Revising data types
    - 4-6 Excluding data before 1950
    
    
5. Exporting wrangled dataframe
** **

## 1. Importing libraries and files
** **


In [32]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [33]:
# Selecting shared path

path = r'C:\Users\mehrn\OneDrive\Desktop\CareerFoundry Materials\Tasks\06 Advanced Analytics\15-03-2023 Canada Wildfires Analysis'

In [34]:
# Importing original open dataset provided by Kaggle from Natural Resources Canada

df_fires = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'CANADA_WILDFIRES.csv'), index_col = 'FID')

### 2. Previewing data, checking columns, and data types  

In [35]:
# Reviewing the number of columns and rows

df_fires.shape

(423831, 8)

In [36]:
# Checking memory usage, columns and their data types in "wildfires" dataframe

df_fires.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 423831 entries, 0 to 423830
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   SRC_AGENCY  423831 non-null  object 
 1   LATITUDE    423831 non-null  float64
 2   LONGITUDE   423831 non-null  float64
 3   REP_DATE    420118 non-null  object 
 4   SIZE_HA     423831 non-null  float64
 5   CAUSE       423590 non-null  object 
 6   PROTZONE    422821 non-null  object 
 7   ECOZ_NAME   423831 non-null  object 
dtypes: float64(3), object(5)
memory usage: 29.1+ MB


In [37]:
# Table preview in "wildfire" dataframe sorted by date and size

df_fires.sort_values(by=['REP_DATE','SIZE_HA']).head(15)

SRC_AGENCY   LATITUDE   LONGITUDE    REP_DATE    SIZE_HA CAUSE  \
FID                                                                     
115345         BC  54.904500 -120.654000  1930-03-07      0.009     H   
399716         YT  61.072500 -138.537500  1946-05-01      0.040     H   
400022         YT  60.983333 -135.033333  1946-05-15      5.300     H   
400023         YT  60.853056 -135.192222  1946-05-20      0.010     H   
400024         YT  60.833611 -135.186667  1946-05-21      0.010     H   
399698         YT  60.937500 -136.285278  1946-06-01   3232.620     H   
399016         YT  61.690795 -134.724170  1946-06-01  37936.040     H   
400681         YT  60.559444 -134.489444  1946-06-05      0.050     H   
399017         YT  62.071944 -136.310556  1946-06-06   2614.500     L   
400025         YT  60.673889 -135.035556  1946-06-14      0.010     H   
399564         YT  61.250000 -138.800000  1946-06-15      0.030     H   
400027         YT  60.522130 -133.367597  1946-06-15   1239.920     H   
400026         YT  60.406918 -134.439634  1946-06-15   4321.950     H   
399814         YT  60.025344 -128.835099  1946-07-10    399.470     H   
419977      PC-WB  59.466600 -113.633300  1946-07-11      3.000     H   

       PROTZONE          ECOZ_NAME  
FID                                 
115345                Boreal PLain  
399716           Boreal Cordillera  
400022           Boreal Cordillera  
400023           Boreal Cordillera  
400024           Boreal Cordillera  
399698           Boreal Cordillera  
399016           Boreal Cordillera  
400681           Boreal Cordillera  
399017           Boreal Cordillera  
400025           Boreal Cordillera  
399564           Boreal Cordillera  
400027           Boreal Cordillera  
400026           Boreal Cordillera  
399814           Boreal Cordillera  
419977                 Taiga Plain

In [38]:
# Descriptive statistics in quantitative columns

df_fires.describe()

LATITUDE      LONGITUDE       SIZE_HA
count  423831.000000  423831.000000  4.238310e+05
mean       51.895936    -104.253547  3.162710e+02
std         4.467770      20.412661  5.617050e+03
min         0.000000    -166.044000  0.000000e+00
25%        49.133700    -120.112350  1.000000e-01
50%        51.023000    -114.503000  1.000000e-01
75%        54.630289     -88.303603  1.000000e+00
max        70.000000     116.188000  1.050000e+06

### 3. Data quality checks


#### 3-1 Duplicates


In [39]:
# Checking for duplicates

df_dups = df_fires[df_fires.duplicated()]
df_dups


SRC_AGENCY   LATITUDE   LONGITUDE    REP_DATE  SIZE_HA CAUSE  \
FID                                                                   
221            BC  54.539000 -128.680000  1951-03-01     0.10     H   
238            BC  54.597000 -128.731000  1951-03-01     0.10     H   
265            BC  54.510000 -128.629000  1953-08-13     0.10     H   
775            BC  53.568000 -120.682000  1950-07-09     0.10     L   
869            BC  50.626000 -117.003000  1950-07-26     0.10     L   
...           ...        ...         ...         ...      ...   ...   
423566      PC-GI   0.000000    0.000000  2013-08-14     0.01     H   
423592      PC-BA   0.000000    0.000000  2015-07-14     0.01     H   
423594      PC-BA   0.000000    0.000000  2015-07-19     0.01     H   
423598      PC-BA   0.000000    0.000000  2015-07-31     0.01     H   
423646      PC-BA  51.387132 -116.129779  2020-07-17     0.01     H   

             PROTZONE           ECOZ_NAME  
FID                                        
221                      Pacific Maritime  
238                      Pacific Maritime  
265                      Pacific Maritime  
775                    Montane Cordillera  
869                    Montane Cordillera  
...               ...                 ...  
423566  Full Response                      
423592  Full Response                      
423594  Full Response                      
423598  Full Response                      
423646  Full Response  Montane Cordillera  

[3230 rows x 8 columns]

#####       
**_Result:_** There are **3230 duplicates** in wildfire reports

#### 3-2 Missing values


In [40]:
# Finding missing values per column in wildfires dataframe

df_fires.isnull().sum()

SRC_AGENCY       0
LATITUDE         0
LONGITUDE        0
REP_DATE      3713
SIZE_HA          0
CAUSE          241
PROTZONE      1010
ECOZ_NAME        0
dtype: int64

#### 
**_Result:_** The **REP_DATE, CAUSE,** and **PROTZONE** columns have missing values in the wildfire reports.

In [41]:
# Reviewing the missing values in REP_DATE

df_null_dates = df_fires[df_fires['REP_DATE'].isnull()== True]
df_null_dates

SRC_AGENCY  LATITUDE  LONGITUDE REP_DATE  SIZE_HA CAUSE       PROTZONE  \
FID                                                                             
100060         BC   58.0833  -122.7000      NaN    2.000     L                  
100072         BC   58.5667  -120.2167      NaN    0.100     L                  
100140         BC   57.8500  -120.2333      NaN    0.000     L                  
100217         BC   55.8833  -120.5000      NaN    0.010     L                  
105111         BC   57.2333  -125.6333      NaN    0.000     L                  
...           ...       ...        ...      ...      ...   ...            ...   
423443      PC-BP   45.2453   -81.5261      NaN    0.001     U  Full Response   
423448      PC-LO   45.9047   -59.9878      NaN   33.300  H-PB                  
423532      PC-LM   42.0000   -72.0000      NaN   28.000  H-PB                  
423534      PC-GF   48.7835  -123.0010      NaN    0.000     H  Full Response   
423535      PC-PU   42.0000   -80.0000      NaN  100.000  H-PB                  

                ECOZ_NAME  
FID                        
100060        Taiga Plain  
100072        Taiga Plain  
100140       Boreal PLain  
100217       Boreal PLain  
105111  Boreal Cordillera  
...                   ...  
423443    MixedWood Plain  
423448  Atlantic Maritime  
423532                     
423534   Pacific Maritime  
423535    MixedWood Plain  

[3713 rows x 8 columns]

#### 
**_Result:_** The **REP_DATE** column has **3713** missing values in wildfire reports which is less than **1%** of data.

In [42]:
# Total number of fires per date 

df_fires['REP_DATE'].value_counts(dropna= False)

NaN           3713
1994-08-03     736
1970-07-16     631
1996-06-12     587
1978-08-04     529
              ... 
1967-11-04       1
1968-11-18       1
1970-11-01       1
1970-11-22       1
2021-03-04       1
Name: REP_DATE, Length: 17665, dtype: int64

In [43]:
# Reviewing the missing values in CAUSE column

df_null_cause = df_fires[df_fires['CAUSE'].isnull() == True]
df_null_cause

SRC_AGENCY   LATITUDE   LONGITUDE    REP_DATE      SIZE_HA CAUSE  \
FID                                                                       
207641         AB  60.014433 -118.498833  2019-06-16       0.0100   NaN   
207701         AB  59.412833 -120.222833  2019-06-25       0.0100   NaN   
207702         AB  59.606867 -117.434033  2019-05-27   74331.6000   NaN   
207768         AB  58.453567 -114.393400  2019-07-21       0.0200   NaN   
207869         AB  58.285283 -117.273167  2019-05-12  350134.8900   NaN   
...           ...        ...         ...         ...          ...   ...   
422901      PC-YO  51.332800 -116.560600  2017-07-03       1.0000   NaN   
423221      PC-WL  49.069200 -113.898100  2019-06-02       0.1000   NaN   
423329      PC-WL  49.020600 -114.044900  2017-07-09       0.0001   NaN   
423371      PC-TN  48.396600  -54.196700  2017-07-16       0.1500   NaN   
423467      PC-PP  41.973800  -82.526400  2017-03-29     147.3000   NaN   

                  PROTZONE           ECOZ_NAME  
FID                                             
207641                             Taiga Plain  
207701                             Taiga Plain  
207702                             Taiga Plain  
207768                            Boreal PLain  
207869                            Boreal PLain  
...                    ...                 ...  
422901       Full Response  Montane Cordillera  
423221  Monitored Response  Montane Cordillera  
423329       Full Response  Montane Cordillera  
423371       Full Response  Boreal Shield East  
423467       Full Response     MixedWood Plain  

[241 rows x 8 columns]

#### 
**_Result:_** The **CAUSE** column has **241** missing values in wildfire reports which is less than **0.05%** of data.

In [44]:
# Finding the missing data in PROTZONE

df_null_protzone = df_fires[df_fires['PROTZONE'].isnull() == True]
df_null_protzone

SRC_AGENCY  LATITUDE  LONGITUDE    REP_DATE  SIZE_HA CAUSE PROTZONE  \
FID                                                                          
238583         MB   54.7546  -101.8559  1971-06-07      0.8     H      NaN   
240461         MB   49.6555   -94.9009  1971-04-17      5.6     H      NaN   
240467         MB   49.6997   -95.1509  1971-08-08      0.2     H      NaN   
241652         MB   53.7361   -93.5787  1977-06-07   1842.4     L      NaN   
243932         MB   50.6660   -94.9993  1974-06-28     16.0     L      NaN   
...           ...       ...        ...         ...      ...   ...      ...   
419686         NT   60.1670  -119.0170  1983-07-15    660.0     L      NaN   
419687         NT   60.0500  -119.3830  1983-06-29   1400.0     L      NaN   
419688         NT   60.0800  -118.7800  1959-07-18   1320.0     L      NaN   
419689         NT   60.0300  -118.7700  1959-07-23    580.0     L      NaN   
419690         NT   60.0000  -119.5000  1960-06-30    455.2     U      NaN   

                 ECOZ_NAME  
FID                         
238583  Boreal Shield West  
240461  Boreal Shield West  
240467  Boreal Shield West  
241652  Boreal Shield West  
243932  Boreal Shield West  
...                    ...  
419686         Taiga Plain  
419687         Taiga Plain  
419688         Taiga Plain  
419689         Taiga Plain  
419690         Taiga Plain  

[1010 rows x 8 columns]

#### 
**_Result:_** The **PROTZONE** column has **1010** missing values in wildfire reports which is less than **0.2%** of data.

#### 3-3 Mixed data types

In [45]:
# Checking data types

df_fires.dtypes

SRC_AGENCY     object
LATITUDE      float64
LONGITUDE     float64
REP_DATE       object
SIZE_HA       float64
CAUSE          object
PROTZONE       object
ECOZ_NAME      object
dtype: object

In [46]:
# Finding the columns with mixed dataypes in wildfires datafram

for col in df_fires.columns.tolist():
    weird = (df_fires[[col]].applymap(type) != df_fires[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_fires[weird]) >0:
        print(col)


REP_DATE
CAUSE
PROTZONE


#### 
**_Result:_** The **REP_DATE, CAUSE,** and **PROTZONE** columns have mixed data type in reported values.

### 4. Data cleaning

####    4-1 Renaming columns' names to keep names clear and consistent


In [47]:
# Renaming index column and the other columns to be consistent and clear

df_fires.rename_axis('fire_id', inplace = True)
df_fires.rename(columns = {'SRC_AGENCY' : 'src_agency'}, inplace = True)
df_fires.rename(columns = {'LATITUDE' : 'latitude'}, inplace = True)
df_fires.rename(columns = {'LONGITUDE' : 'longitude'}, inplace = True)
df_fires.rename(columns = {'REP_DATE' : 'rep_date'}, inplace = True)
df_fires.rename(columns = {'SIZE_HA' : 'size'}, inplace = True)
df_fires.rename(columns = {'CAUSE' : 'cause'}, inplace = True)
df_fires.rename(columns = {'PROTZONE' : 'prot_zone'}, inplace = True)
df_fires.rename(columns = {'ECOZ_NAME' : 'eco_name'}, inplace = True)
df_fires

src_agency  latitude   longitude    rep_date        size cause  \
fire_id                                                                  
0               BC  59.96300 -128.172000  1953-05-26      8.0000     H   
1               BC  59.31800 -132.172000  1950-06-22      8.0000     L   
2               BC  59.87600 -131.922000  1950-06-04  12949.9000     H   
3               BC  59.76000 -132.808000  1951-07-15    241.1000     H   
4               BC  59.43400 -126.172000  1952-06-12      1.2000     H   
...            ...       ...         ...         ...         ...   ...   
423826       PC-WB  59.61015 -113.985117  2021-07-08     65.1000     L   
423827       PC-WL  49.11120 -113.836000  2021-03-04      7.6000  H-PB   
423828       PC-WL  49.11220 -113.840500  2021-03-17      2.9000  H-PB   
423829       PC-WL  49.04630 -113.916300  2021-09-18      0.0001     H   
423830       PC-WP  57.84200  -93.788000  2021-07-12      3.0000     L   

                  prot_zone            eco_name  
fire_id                                          
0                             Boreal Cordillera  
1                             Boreal Cordillera  
2                             Boreal Cordillera  
3                             Boreal Cordillera  
4                             Boreal Cordillera  
...                     ...                 ...  
423826   Monitored Response         Taiga Plain  
423827      Prescribed Fire  Montane Cordillera  
423828      Prescribed Fire  Montane Cordillera  
423829        Full Response  Montane Cordillera  
423830   Monitored Response        Hudson Plain  

[423831 rows x 8 columns]

#### 4-2 Missing values treatment

In [48]:
# Dropping missing value as they are ~1% of data 

df_fires.dropna(inplace= True)
df_fires

src_agency  latitude   longitude    rep_date        size cause  \
fire_id                                                                  
0               BC  59.96300 -128.172000  1953-05-26      8.0000     H   
1               BC  59.31800 -132.172000  1950-06-22      8.0000     L   
2               BC  59.87600 -131.922000  1950-06-04  12949.9000     H   
3               BC  59.76000 -132.808000  1951-07-15    241.1000     H   
4               BC  59.43400 -126.172000  1952-06-12      1.2000     H   
...            ...       ...         ...         ...         ...   ...   
423826       PC-WB  59.61015 -113.985117  2021-07-08     65.1000     L   
423827       PC-WL  49.11120 -113.836000  2021-03-04      7.6000  H-PB   
423828       PC-WL  49.11220 -113.840500  2021-03-17      2.9000  H-PB   
423829       PC-WL  49.04630 -113.916300  2021-09-18      0.0001     H   
423830       PC-WP  57.84200  -93.788000  2021-07-12      3.0000     L   

                  prot_zone            eco_name  
fire_id                                          
0                             Boreal Cordillera  
1                             Boreal Cordillera  
2                             Boreal Cordillera  
3                             Boreal Cordillera  
4                             Boreal Cordillera  
...                     ...                 ...  
423826   Monitored Response         Taiga Plain  
423827      Prescribed Fire  Montane Cordillera  
423828      Prescribed Fire  Montane Cordillera  
423829        Full Response  Montane Cordillera  
423830   Monitored Response        Hudson Plain  

[418867 rows x 8 columns]

#### 4-3 Removing duplicates


In [49]:
# Dropping duplicates

df_fires.drop_duplicates()

src_agency  latitude   longitude    rep_date        size cause  \
fire_id                                                                  
0               BC  59.96300 -128.172000  1953-05-26      8.0000     H   
1               BC  59.31800 -132.172000  1950-06-22      8.0000     L   
2               BC  59.87600 -131.922000  1950-06-04  12949.9000     H   
3               BC  59.76000 -132.808000  1951-07-15    241.1000     H   
4               BC  59.43400 -126.172000  1952-06-12      1.2000     H   
...            ...       ...         ...         ...         ...   ...   
423826       PC-WB  59.61015 -113.985117  2021-07-08     65.1000     L   
423827       PC-WL  49.11120 -113.836000  2021-03-04      7.6000  H-PB   
423828       PC-WL  49.11220 -113.840500  2021-03-17      2.9000  H-PB   
423829       PC-WL  49.04630 -113.916300  2021-09-18      0.0001     H   
423830       PC-WP  57.84200  -93.788000  2021-07-12      3.0000     L   

                  prot_zone            eco_name  
fire_id                                          
0                             Boreal Cordillera  
1                             Boreal Cordillera  
2                             Boreal Cordillera  
3                             Boreal Cordillera  
4                             Boreal Cordillera  
...                     ...                 ...  
423826   Monitored Response         Taiga Plain  
423827      Prescribed Fire  Montane Cordillera  
423828      Prescribed Fire  Montane Cordillera  
423829        Full Response  Montane Cordillera  
423830   Monitored Response        Hudson Plain  

[415649 rows x 8 columns]

#### 4-4 Dropping unnecessary columns

We decided to keep all the columns for now.

#### 4-5 Revising data types

In [50]:
# Revising data types of mixed type columns

df_fires['src_agency'] = df_fires['src_agency'].astype('str')
df_fires['prot_zone'] = df_fires['prot_zone'].astype('str')
df_fires['cause'] = df_fires['cause'].astype('str')

In [51]:
# Double checking data type of columns

df_fires.dtypes

src_agency     object
latitude      float64
longitude     float64
rep_date       object
size          float64
cause          object
prot_zone      object
eco_name       object
dtype: object

In [52]:
# Counting the number of distinct elements in each column

df_fires.nunique()

src_agency        56
latitude      141140
longitude     208624
rep_date       17660
size           14870
cause              5
prot_zone         40
eco_name          17
dtype: int64

In [53]:
# Exploring data by grouping it using agency and date

df_fires.groupby(['src_agency', 'rep_date']).agg({'src_agency': ['size']})

src_agency
                            size
src_agency rep_date             
AB         1959-04-27          1
           1959-05-14          2
           1959-05-15          2
           1959-05-16          2
           1959-05-17          2
...                          ...
YT         2021-08-01          2
           2021-08-02          1
           2021-08-03          3
           2021-08-05          2
           2021-08-06          3

[71431 rows x 1 columns]

#### 4-6 Excluding data before 1950

In [54]:
# Excluding rows before year 1950

df = df_fires[df_fires['rep_date'] >= '1950-01-01']
df

src_agency  latitude   longitude    rep_date        size cause  \
fire_id                                                                  
0               BC  59.96300 -128.172000  1953-05-26      8.0000     H   
1               BC  59.31800 -132.172000  1950-06-22      8.0000     L   
2               BC  59.87600 -131.922000  1950-06-04  12949.9000     H   
3               BC  59.76000 -132.808000  1951-07-15    241.1000     H   
4               BC  59.43400 -126.172000  1952-06-12      1.2000     H   
...            ...       ...         ...         ...         ...   ...   
423826       PC-WB  59.61015 -113.985117  2021-07-08     65.1000     L   
423827       PC-WL  49.11120 -113.836000  2021-03-04      7.6000  H-PB   
423828       PC-WL  49.11220 -113.840500  2021-03-17      2.9000  H-PB   
423829       PC-WL  49.04630 -113.916300  2021-09-18      0.0001     H   
423830       PC-WP  57.84200  -93.788000  2021-07-12      3.0000     L   

                  prot_zone            eco_name  
fire_id                                          
0                             Boreal Cordillera  
1                             Boreal Cordillera  
2                             Boreal Cordillera  
3                             Boreal Cordillera  
4                             Boreal Cordillera  
...                     ...                 ...  
423826   Monitored Response         Taiga Plain  
423827      Prescribed Fire  Montane Cordillera  
423828      Prescribed Fire  Montane Cordillera  
423829        Full Response  Montane Cordillera  
423830   Monitored Response        Hudson Plain  

[418791 rows x 8 columns]

### 5. Exporting wrangled dataframe


In [55]:
# Exporting wrangled datafram in PKL format

df.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'wrangled_wildfires.pkl'))